### PR TITLE
Completion: Suggest types after keywords indicating a local function

### LINKED ISSUE #53585: Completion doesn't offer return types for static local functions
**Version Used**: 

16.9.4

**Steps to Reproduce**:

image

**Expected Behavior**:

Types are offered.

**Actual Behavior**:

I have to type the return type I want manually. 
image

### PR DESCRIPTION
**Pull request by MaStr11**
Fixes #53585

cc @Youssef1313

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ private ImmutableArray<ISymbol> GetSymbolsForCurrentContext()
                 // is a type-only context, we'll show all symbols anyway.
                 return GetSymbolsForExpressionOrStatementContext();
             }
-            else if (_context.IsTypeContext || _context.IsNamespaceContext)
+            else if (_context.IsTypeContext || _context.IsNamespaceContext || _context.IsLocalFunctionDeclarationContext)
```
Comment (MaStr11) [inline src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs:81]: @CyrusNajmabadi This is probably naive or just plain wrong.  Do you have any objections? If not, I will add some more tests.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 2]
```diff
@@ -2168,6 +2168,24 @@ void M(String parameter)
...
+using System;
+class C
+{
+    void M(String parameter)
+    {
+        static $$
+    }
+}
+";
+            await VerifyItemExistsAsync(markup, "String");
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:2185]: Is `string` offered too? It has a separate keyword recommender IIRC (there are other keywords too).

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs (hunk 1 above)
Comment (MaStr11) [reply src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs]: I gave this a try in 842a1f6 and removed the new property in eb1b5be. This results in one test failure, discussed below.

**Comment by MaStr11** [comment]
@CyrusNajmabadi This simple approach expands `syntaxTree.IsTypeContext`, to also be true in `syntaxTree.IsLocalFunctionDeclarationContext`. The only test that fails with this addition is the `delegate` keyword recommender (test failure: `delegate` is suggested after `static`):

https://github.com/dotnet/roslyn/blob/6453b4f67c612dd68fb486d56a39eac88545e340/src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs#L289-L291

The root cause is here in `ValidTypeContext`:
https://github.com/dotnet/roslyn/blob/6453b4f67c612dd68fb486d56a39eac88545e340/src/Features/CSharp/Portable/Completion/KeywordRecommenders/DelegateKeywordRecommender.cs#L31-L47

I'm not too confident about the `delegate` keyword, but to me, it feels like it should *not* be suggested in a type context. For instance, it is suggested in method parameters (`void M($$`), which I think is wrong. @CyrusNajmabadi It probably should read `context.IsTypeDeclarationContext` in line 44?

@CyrusNajmabadi What do you think about the approach of expanding `syntaxTree.IsTypeContext`? What should we do about the `DelegateKeywordRecommender`.

**Comment by CyrusNajmabadi** [comment]
`delegate` is now legal in a type-context due to function pointers (which look like `delegate*<int, void>`).  So i would update thigns accordingly based on that.  If you still have any remaining issues/questions, lmk at that point.  Thanks!

File: src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs (hunk 1 above)
Comment (MaStr11) [reply src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs]: I tracked down every reference of `IsTypeContext` (at least I tried) and added tests for each occurrence.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 2 above)
Comment (MaStr11) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs:2185]: It wasn't offered. I fixed this and added tests.

**Comment by MaStr11** [comment]
@CyrusNajmabadi Ready for review

File: src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs [hunk 3]
```diff
@@ -288,7 +288,7 @@ public async Task TestNotAfterSealed()
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestAfterStatic()
         {
-            await VerifyAbsenceAsync(SourceCodeKind.Regular, @"static $$");
+            await VerifyKeywordAsync(SourceCodeKind.Regular, @"static $$");
             await VerifyKeywordAsync(SourceCodeKind.Script, @"static $$");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:292]: can we have a test for this after static inside a method as well?

File: src/EditorFeatures/CSharpTest2/Recommendations/UShortKeywordRecommenderTests.cs [hunk 4]
```diff
@@ -761,5 +761,13 @@ class C
 {
     delegate*$$");
         }
+
+        [WorkItem(53585, "https://github.com/dotnet/roslyn/issues/53585")]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        public async Task TestAfterStaticLocalFunction()
+        {
+            await VerifyKeywordAsync(AddInsideMethod(@"
+static $$"));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/UShortKeywordRecommenderTests.cs:770]: so, you had all the tests after 'static' in a method, can you do the same for after 'async'? as well?  or make all the tests theories and test `static` `async` `async static` and `static async`.  Thanks!

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/IntKeywordRecommender.cs [hunk 5]
```diff
@@ -37,6 +37,7 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
                 context.IsFixedVariableDeclarationContext ||
                 context.IsParameterTypeContext ||
                 context.IsPossibleLambdaOrAnonymousMethodParameterTypeContext ||
+                context.IsLocalFunctionDeclarationContext ||
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/IntKeywordRecommender.cs:40]: i think this is proably better named as `IsLocalFunctionReturnTypeContext` (similar to `IsDelegateReturnTypeContext`).

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 6]
```diff
@@ -353,6 +361,12 @@ static bool IsGlobalAttributeList(AttributeListSyntax attributeList)
             return false;
         }
 
+        public static bool IsLocalFunctionDeclarationContext(
+            this SyntaxTree syntaxTree,
+            int position,
+            CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:367]:
```suggestion
        public static bool IsLocalFunctionDeclarationContext(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/IntKeywordRecommender.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/IntKeywordRecommender.cs:40]: n/m.  i see this name already exists.

**Review by CyrusNajmabadi — APPROVED** [review]
LGTM with requested test additions.

File: src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs (hunk 3 above)
Comment (MaStr11) [reply src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:292]: Done in b581ce2ac282cc8592047a134758e4d24007955a (with a fix in 8439d03)

File: src/EditorFeatures/CSharpTest2/Recommendations/UShortKeywordRecommenderTests.cs (hunk 4 above)
Comment (MaStr11) [reply src/EditorFeatures/CSharpTest2/Recommendations/UShortKeywordRecommenderTests.cs]: I used a `ClassData` theory added in b46098a and applied in f686b67.

**Comment by MaStr11** [comment]
@CyrusNajmabadi Feedback is addressed and tests are green (except one unrelated integration test failure). I added the new file `TheoryDataKeywordsIndicatingLocalFunction.cs` for the theory data (`[ClassData(typeof(...)]`) for reuse in all the recommender tests. I hope that's fine (using `[MemberData]` and a property on the tests base class would be an alternative).

https://github.com/dotnet/roslyn/blob/8439d0366cbe31643f4e9ad6857e9ff047187195/src/EditorFeatures/CSharpTest2/Recommendations/TheoryDataKeywordsIndicatingLocalFunction.cs#L11-L16

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 7]
```diff
@@ -353,6 +361,9 @@ static bool IsGlobalAttributeList(AttributeListSyntax attributeList)
             return false;
         }
 
+        public static bool IsLocalFunctionDeclarationContext(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        => syntaxTree.IsLocalFunctionDeclarationContext(position, s_validLocalFunctionModifiers, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:365]:
```suggestion
            => syntaxTree.IsLocalFunctionDeclarationContext(position, s_validLocalFunctionModifiers, cancellationToken);
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**